File: src/shared/secrets.mjs
// Shared Parameter Store credential loader for Node.js (Lambda) code.
//
// Replaces the legacy hardcoded credentials. Reads SecureString parameters
// from SSM under /sb/<env>/<group>/<key>, with a small in-process cache so
// hot Lambdas don't re-fetch on every invocation.
//
// The AWS SDK v3 is bundled into the Lambda Node.js runtime, so this adds no
// deployment dependency.
//
// Usage:
//   import { getSecret, getGroup } from './shared/secrets.mjs';
//   const apiKey = await getSecret('orderdesk', 'api-key');
//   const ftp = await getGroup('ftp'); // { host, user, password }

import { SSMClient, GetParameterCommand, GetParametersByPathCommand } from '@aws-sdk/client-ssm';

const REGION = process.env.AWS_REGION || process.env.AWS_DEFAULT_REGION || 'us-east-1';
const ENV = process.env.SB_ENV || process.env.NODE_ENV_NAME || 'dev';
const TTL_MS = Number(process.env.SECRET_CACHE_TTL_MS || 5 * 60 * 1000); // 5 min

const client = new SSMClient({ region: REGION });
const cache = new Map(); // name -> { value, expires }

function prefix() {
  return `/sb/${ENV}`;
}

function cached(name) {
  const hit = cache.get(name);
  if (hit && hit.expires > Date.now()) return hit.value;
  return undefined;
}

function store(name, value) {
  cache.set(name, { value, expires: Date.now() + TTL_MS });
  return value;
}

/**
 * Fetch a single secret value by group + key, e.g. getSecret('orderdesk', 'api-key').
 * Returns the decrypted string. Throws if the parameter is missing.
 */
export async function getSecret(group, key) {
  const name = `${prefix()}/${group}/${key}`;
  const hit = cached(name);
  if (hit !== undefined) return hit;

  const res = await client.send(
    new GetParameterCommand({ Name: name, WithDecryption: true }),
  );
  return store(name, res.Parameter?.Value);
}

/**
 * Fetch all keys within a group as an object, e.g. getGroup('ftp')
 * -> { host, user, password }. Uses one GetParametersByPath call.
 */
export async function getGroup(group) {
  const path = `${prefix()}/${group}`;
  const hit = cached(path);
  if (hit !== undefined) return hit;

  const out = {};
  let nextToken;
  do {
    const res = await client.send(
      new GetParametersByPathCommand({
        Path: path,
        Recursive: true,
        WithDecryption: true,
        NextToken: nextToken,
      }),
    );
    for (const p of res.Parameters || []) {
      const leaf = p.Name.slice(p.Name.lastIndexOf('/') + 1);
      out[leaf] = p.Value;
    }
    nextToken = res.NextToken;
  } while (nextToken);

  return store(path, out);
}

/** Clear the in-process cache (useful in tests). */
export function clearSecretCache() {
  cache.clear();
}

File: src/shared/secrets.py
"""Shared Parameter Store credential loader for Python (ECS Fargate) code.

Replaces the legacy hardcoded credentials. Reads SecureString parameters from
SSM under /sb/<env>/<group>/<key>, with a small in-process TTL cache so
long-running services don't re-fetch on every job.

Usage:
    from shared.secrets import get_secret, get_group
    api_key = get_secret("orderdesk", "api-key")
    ftp = get_group("ftp")  # {"host": ..., "user": ..., "password": ...}
"""

from __future__ import annotations

import os
import time
from typing import Dict, Optional

import boto3

_REGION = os.environ.get("AWS_REGION") or os.environ.get("AWS_DEFAULT_REGION") or "us-east-1"
_ENV = os.environ.get("SB_ENV", "dev")
_TTL_S = float(os.environ.get("SECRET_CACHE_TTL_S", 5 * 60))  # 5 min

_client = boto3.client("ssm", region_name=_REGION)
_cache: Dict[str, tuple[float, object]] = {}  # name -> (expires, value)


def _prefix() -> str:
    return f"/sb/{_ENV}"


def _cached(name: str):
    hit = _cache.get(name)
    if hit and hit[0] > time.time():
        return hit[1]
    return None


def _store(name: str, value):
    _cache[name] = (time.time() + _TTL_S, value)
    return value


def get_secret(group: str, key: str) -> Optional[str]:
    """Fetch a single decrypted secret value by group + key."""
    name = f"{_prefix()}/{group}/{key}"
    hit = _cached(name)
    if hit is not None:
        return hit  # type: ignore[return-value]

    resp = _client.get_parameter(Name=name, WithDecryption=True)
    return _store(name, resp["Parameter"]["Value"])


def get_group(group: str) -> Dict[str, str]:
    """Fetch all keys within a group as a dict, via one paginated call."""
    path = f"{_prefix()}/{group}"
    hit = _cached(path)
    if hit is not None:
        return hit  # type: ignore[return-value]

    out: Dict[str, str] = {}
    paginator = _client.get_paginator("get_parameters_by_path")
    for page in paginator.paginate(Path=path, Recursive=True, WithDecryption=True):
        for p in page.get("Parameters", []):
            leaf = p["Name"].rsplit("/", 1)[-1]
            out[leaf] = p["Value"]

    return _store(path, out)


def clear_secret_cache() -> None:
    """Clear the in-process cache (useful in tests)."""
    _cache.clear()
